// term-logger/dist/structures/BgColors.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BgColors = void 0;
/**
 * Defines various text backgrounds using chalk library.
 */
const chalk_1 = __importDefault(require("chalk"));
/**
 * An object that represents different text backgrounds.
 */
const BACKGROUND_COLORS = {
    blue: "bgBlue",
    cyan: "bgCyan",
    gray: "bgGray",
    green: "bgGreen",
    purple: "bgMagenta",
    red: "bgRed",
    white: "bgWhite",
    yellow: "bgYellow",
};
/**
 * Object that maps background color names to functions that log a message with the specified background color.
 */
exports.BgColors = Object.fromEntries(Object.entries(BACKGROUND_COLORS).map(([key, value]) => [
    key,
    /**
     * Logs a message with the specified background color.
     * @param message - The message to be logged.
     */
    (message) => console.log(chalk_1.default[value](message)),
]));
